### ISSUE TITLE
Refetch triggers twice

### ISSUE THREAD

**Issue by peterpetre** [apollographql/apollo-client#3948]
**Intended outcome:**
`refetch` should always work correctly

**Actual outcome:**
`refetch` sometimes triggers twice, one with the new variables (which is correct) and after that with the initial variables

**How to reproduce the issue:**
i can not reproduce constantly, sometimes works ok, sometimes does not, i tried many things from using the old HOCs `graphql` to new react render-prop components, changing fetchPolicy and more

**Versions**
latest

**Comment by peterpetre** [comment]
maybe there is something with how things are handled on `isSetVariables = true`, `previousVariables = previousQuery.variables`, `networkStatus = NetworkStatus.setVariables` and `networkStatus = NetworkStatus.refetch`?

**Comment by peterpetre** [comment]
same goes for fetchMore

**Comment by pl12133** [comment]
Reproduction: https://codesandbox.io/s/yvk8n4j4lx

This is probably a duplicate of #2285

**Comment by peterpetre** [comment]
FYI it was happening with using graphql hoc too, so i am not sure that this is just a react Query component issue. Btw the issue was too serious for our company's product, imagine our users sometimes couldnt use filter, search and load more... so we just dropped apollo client and create our own version. Good luck!

**Comment by peterpetre** [comment]
The issue was too serious for our company's product, imagine our users sometimes couldnt use filter, search and load more... so we dropped apollo client and created our own version. Good luck!

**Comment by danilobuerger** [comment]
Closing as duplicate of #2285

**Comment by JamieKudla** [comment]
This is still an issue, but the "OP" was just closed #2285

**Comment by rossm6** [comment]
This might help somebody.

I was using pagination but not the cache with useLazyQuery.  The first fetch worked fine but my subsequent calls to fetchMore would show a successful query per the network tab but the UI would like it hasn't updated because a second query was being made with the original variables.  I'm quite new to React and this caught me out I think.

Originally I had something like this in the parent component -

```
function Parent () {
  const [fetchInitialData, { loading, data, fetchMore, called }] = useLazyQuery(LOAD_PRODUCTS, { fetchPolicy: 'no-cache' });
  
  const getData = React.useCallback(() => { // call fetchMore in here}, 
   [called, fetchMore, fetchInitialData]);

  render (
    <Child fetchData={getData} />
  )
```

}


The problem I think with this is that React doesn't notice any difference between the fetchMore after the first call.  The solution is to avoid React.useCallback altogether and instead use this pattern -

```
function Parent () {
  const [fetchInitialData, { loading, data, fetchMore, called }] = useLazyQuery(LOAD_PRODUCTS, { fetchPolicy: 'no-cache' });
  
  const privateFunction = () => { // call fetchMore };

  const getData = useRef((o) => { return privateFunction(o) });

  render (
    <Child fetchData={getData.current} />
  )

}
```

This does the same thing as useCallback in that with have a fixed ref which can be passed to the child component and included in dependency arrays but at the same time the function call will use the latest fetchMore function object.
